tests/scenario/test_charm.py

```diff
@@ -0,0 +1,482 @@
+    )
+    state = scenario.State(relations=[relation])
+    out = ctx.run_action("get-outstanding-certificate-requests", state)
+    assert not out.success
+    assert out.failure == "Failed to parse outstanding requests"
+
+
+def test_given_requirer_application_when_get_outstanding_certificate_requests_action_then_csrs_information_is_returned():  # noqa: E501
+    example_unit_csrs = [
+        {
+            "relation_id": 1234,
+            # SCENARIO-NOTE: in the Harness test, this is 'unit/0', but I don't
+            # think you can actually name units with Juju?
+            "unit_name": "application/0",
+            "application_name": "application",
+            "unit_csrs": [{"certificate_signing_request": "some csr"}],
+        }
+    ]
+    ctx = scenario.Context(ManualTLSCertificatesCharm)
+    # SCENARIO-NOTE: converting this test was a little tricky, because the
+    # Harness test doesn't set up the relation data, it mocks the return value
+    # for a method that processes the relation data. Harness and Scenario can
+    # equally do this by setting up the relation data, but it seems like the
+    # mocking approach doesn't suit Scenario - is that an issue?
+    relation = scenario.Relation(
+        # SCENARIO-NOTE: this is in the Harness test, but it seems like it could
+        # be left out if the ID was automatic.
+        relation_id=1234,
+        # SCENARIO-NOTE: 'endpoint' is an unfamiliar name to me in this context.
+        endpoint="certificates",
+        remote_app_name="application",
+        remote_units_data={
+            0: {
+                "certificate_signing_requests": json.dumps(
+                    [{"certificate_signing_request": "some csr"}]
+                )
+            }
+        },
+    )
+    state = scenario.State(relations=[relation])
+    out = ctx.run_action("get-outstanding-certificate-requests", state)
+    assert out.success
+    assert json.loads(out.results["result"]) == example_unit_csrs
+
+
+def test_given_requirer_and_no_outstanding_certs_when_get_outstanding_certificate_requests_action_then_empty_list_is_returned():  # noqa: E501
+    ctx = scenario.Context(ManualTLSCertificatesCharm)
+    relation = scenario.Relation(
+        # SCENARIO-NOTE: 'endpoint' is an unfamiliar name to me in this context.
+        endpoint="certificates",
+        remote_app_name="requirer",
+        remote_units_data={0: {"certificate_signing_requests": json.dumps([])}},
+    )
+    state = scenario.State(relations=[relation])
+    out = ctx.run_action("get-outstanding-certificate-requests", state)
+    assert out.success
+    assert json.loads(out.results["result"]) == []
+
+
+def test_given_relation_id_not_exist_when_get_outstanding_certificate_requests_action_then_action_returns_empty_list():  # noqa: E501
+    ctx = scenario.Context(ManualTLSCertificatesCharm)
+    relation = scenario.Relation(
+        # SCENARIO-NOTE: 'endpoint' is an unfamiliar name to me in this context.
```

Comment:
Which was my ignorance, or I had forgotten the term, maybe. If I look up "integrations" then it's clearly there.